x/erc20/keeper/ibc_callbacks.go

```diff
@@ -82,41 +68,59 @@ func (k Keeper) OnRecvPacket(
 		data.Denom, data.Amount,
 	)
 
-	// check if the coin is a native staking token
-	bondDenom := k.stakingKeeper.BondDenom(ctx)
-	if coin.Denom == bondDenom {
-		// no-op, received coin is the staking denomination
+	pairID := k.GetTokenPairID(ctx, coin.Denom)
+	pair, _ := k.GetTokenPair(ctx, pairID)
+	// Case 1 - Coin is native EVMOS
+	if !pair.Enabled || pair.IsNativeCoin() {
+		// no-op: continue with the rest of the stack without registration
 		return ack
 	}
 
-	pairID := k.GetTokenPairID(ctx, coin.Denom)
-	if len(pairID) == 0 {
-		// short-circuit: if the denom is not registered, conversion will fail
-		// so we can continue with the rest of the stack
-		return ack
+	// Case 2 - native ERC20 token
+	if pair.IsNativeERC20() {
+		// ERC20 module or token pair is disabled -> return
+		if !k.IsERC20Enabled(ctx) || !pair.Enabled {
+			return ack
+		}
+
+		msgConvert := types.NewMsgConvertERC20(coin.Amount, receiver, pair.GetERC20Contract(), common.BytesToAddress(sender))
+		// Convert from Coin to ERC20
+		_, err := k.ConvertERC20(ctx, msgConvert)
+		if err != nil {
+			return channeltypes.NewErrorAcknowledgement(err)
+		}
 	}
 
-	pair, _ := k.GetTokenPair(ctx, pairID)
-	if !pair.Enabled {
-		// no-op: continue with the rest of the stack without conversion
+	// Case 3 - IBC Voucher
+	// short-circuit: if the denom is not a single hop voucher
+	// TODO: Consider how it integrates with PFM.
+	if !ibc.IsSingleHop(data.Denom) {
 		return ack
 	}
 
-	// Instead of converting just the received coins, convert the whole user balance
-	// which includes the received coins.
-	balance := k.bankKeeper.GetBalance(ctx, recipient, coin.Denom)
-
-	// Build MsgConvertCoin, from recipient to recipient since IBC transfer already occurred
-	msg := types.NewMsgConvertCoin(balance, common.BytesToAddress(recipient.Bytes()), recipient)
-
-	// NOTE: we don't use ValidateBasic the msg since we've already validated
-	// the ICS20 packet data
-
-	// Use MsgConvertCoin to convert the Cosmos Coin to an ERC20
-	if _, err = k.ConvertCoin(sdk.WrapSDKContext(ctx), msg); err != nil {
+	denomAddr, err := utils.GetIBCDenomAddress(coin.Denom)
+	if err != nil {
 		return channeltypes.NewErrorAcknowledgement(err)
 	}
 
+	// Truncate to 20 bytes (40 hex characters)
+	truncatedAddr := denomAddr[:20]
+	params := k.evmKeeper.GetParams(ctx)
+	found := params.IsPrecompileRegistered(common.BytesToAddress(truncatedAddr).String())
+	if !found {
+		// TODO: Should we use the new RegisterCoin method here instead?
```

Comment:
what does the Register coin method do differently?

Reply:
Just move the registration of the precompile and the update of the token pair to that method separately ?

Reply:
ah then yes, let's use that function instead